python/OnnxUdf/onnx_udf.py

```diff
@@ -0,0 +1,76 @@
+import functools
+import sys
+from typing import Dict
+import xarray as xr
+import numpy as np
+from openeo.udf.debug import inspect
+
+# The onnx_deps folder contains the extracted contents of the dependencies archive provided in the job options
+sys.path.insert(0, "onnx_deps") 
+import onnxruntime as ort
+
+@functools.lru_cache(maxsize=5)
+def _load_ort_session(model_name: str) -> ort.InferenceSession:
+    """
+    Loads an ONNX model from the onnx_models folder and returns an ONNX runtime session.
+
+    Extracting the model loading code into a separate function allows us to cache the loaded model.
+    This prevents the model from being loaded for every chunk of data that is processed, but only once per executor.
+
+    Should you have to download the model from a remote location, you can add the download code here, and cache the model.
+
+    Make sure that the arguments of the method you add the @functools.lru_cache decorator to are hashable.
+    Be careful with using this decorator for class methods, as the self argument is not hashable. 
+    In that case, you can use a static method.
+
+    More information on this functool can be found here: 
+    https://docs.python.org/3/library/functools.html#functools.lru_cache
+    """
+    # The onnx_models folder contains the content of the model archive provided in the job options
+    return ort.InferenceSession(f"onnx_models/{model_name}")
+
+def _apply_model(input_xr: xr.DataArray) -> xr.DataArray:
+    """
+    Run the inference on the given input data using the provided ONNX runtime session.
+    This method is called for each timestep in the chunk received by apply_datacube.
+    """
+    # Load the ONNX model
+    ort_session = _load_ort_session("test_model.onnx") # name of the model in the archive
+
+    # Make sure the input dimensions are in the expected order and save the original shape
+    input_xr = input_xr.transpose("bands", "y", "x")
+    input_shape = input_xr.shape
+
+    # Get the underlying np.ndarray and reshape it to the expected shape.
+    input_np = input_xr.values.reshape(ort_session.get_inputs()[0].shape) 
```

Comment:
is it possible to first reshape using xarray and only afterwards get out the numpy array?

That might further emphasize the use of xarray for all you operations

Reply:
It is possible, however the only reason a reshape is needed is to add a diminsion (1,256,256 -> 1,1,256,256). Adding a dimension in xarray seems more difficult as you need to name it. This approach seemed more straightforward, while the transpose step is still shown.